components/renku_data_services/data_connectors/db.py

```diff
@@ -280,24 +326,66 @@ async def insert_data_connector(
             created_by_id=user.id,
             description=data_connector.description,
             keywords=data_connector.keywords,
+            global_slug=slug if isinstance(data_connector, models.UnsavedGlobalDataConnector) else None,
         )
-        data_connector_slug = ns_schemas.EntitySlugORM.create_data_connector_slug(
-            slug,
-            data_connector_id=data_connector_orm.id,
-            namespace_id=ns.id,
-            project_id=project.id if project else None,
-        )
+        if ns is not None:
+            data_connector_slug = ns_schemas.EntitySlugORM.create_data_connector_slug(
+                slug,
+                data_connector_id=data_connector_orm.id,
+                namespace_id=ns.id,
+                project_id=project.id if project else None,
+            )
 
         session.add(data_connector_orm)
-        session.add(data_connector_slug)
+        if ns is not None:
+            session.add(data_connector_slug)
         await session.flush()
         await session.refresh(data_connector_orm)
-        await session.refresh(data_connector_slug)
+        if ns is not None:
+            await session.refresh(data_connector_slug)
         if project:
             await session.refresh(data_connector_slug.project)
 
         return data_connector_orm.dump()
 
+    @with_db_transaction
+    async def insert_global_data_connector(
+        self,
+        user: base_models.APIUser,
+        data_connector: models.UnsavedGlobalDataConnector,
+        validator: RCloneValidator | None,
+        *,
+        session: AsyncSession | None = None,
+    ) -> tuple[models.GlobalDataConnector, bool]:
+        """Insert a new global data connector entry."""
+        if not session:
+            raise errors.ProgrammingError(message="A database session is required.")
+        if user.id is None:
+            raise errors.UnauthorizedError(message="You do not have the required permissions for this operation.")
+
+        slug = data_connector.slug or base_models.Slug.from_name(data_connector.name).value
+
+        existing_global_dc_stmt = select(schemas.DataConnectorORM).where(schemas.DataConnectorORM.global_slug == slug)
+        existing_global_dc = await session.scalar(existing_global_dc_stmt)
+        if existing_global_dc is not None:
+            dc = existing_global_dc.dump()
+            if not isinstance(dc, models.GlobalDataConnector):
+                raise errors.ProgrammingError(message=f"Expected to get a global data connector ('{dc.id}')")
+            return dc, False
+
+        # Fully validate a global data connector before inserting
+        if isinstance(data_connector, models.UnsavedGlobalDataConnector):
+            if validator is None:
+                raise RuntimeError("Could not validate global data connector")
+            data_connector = await validate_unsaved_global_data_connector(
+                data_connector=data_connector, validator=validator
```

Comment:
suggestion: I think we should allow only updating the name, description, keywords and maybe storage on global data connectors. For trying to change anything else we should fail with NotImplementedError. We can make a check for these conditions right at the top of this method and if things we do not expect to change are being changed then we fail. If not I think we can just fallback and continue with the old code and it will work for both global and non-global data connectors.

Reply:
Why is this comment on the `insert` method? Not sure what you mean here.

Reply:
I have no clue why are things so shifted. I think that there were commits being pushed as I was doing the review. It was meant to be on the update_data_connector method in db.py

Reply:
> The main thing that I find concerning is that the global data connectors slug have their slugs in the data connectors table and not in the other tables where all other slugs are stored.

I am much much more in favour of doing the update only on the data connectors table rather than messing up the entity slug table which would have consequences for projects too.

Reply:
The patch validation will reject `namespace` and `slug` updates on global data connectors for now.